### PR TITLE
Add ability to create temp mapped drive for integration tests

### LINKED ISSUE #7330: Allow MSBuild tests to access root of drive
### Issue Description
Allow drive enumerating tests to access the root of a drive inside of Visual Studio. Currently, an `UnauthorizedAccessException` is thrown (so drive substitution is required for such tests to pass in VS).

### Steps to Reproduce
Run `subst z: <drive>:\<path>` (or preferably, `subst z: %TEMP%`) on the command line for the below test to pass. This test can be added inside of CreateItem_Tests.cs.
 ```
        [Fact]
        public void WildcardDriveEnumerationTaskItemLogsWarning()
        {
            using (var env = TestEnvironment.Create())
            {
                ChangeWaves.ResetStateForTests();
                env.SetEnvironmentVariable("MSBUILDFAILONDRIVEENUMERATINGWILDCARD", "0");
                BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();

                CreateItem t = new CreateItem();
                MockEngine engine = new MockEngine();

                t.BuildEngine = engine;
                t.Include = new ITaskItem[] { new TaskItem(@"z:\**\*.log") };
                bool succeeded = t.Execute();

                Assert.True(succeeded);
                Assert.Equal(1, engine.Warnings);
                engine.AssertLogContains("results in drive enumeration");

                ChangeWaves.ResetStateForTests();
            }
        }
```

### Expected Behavior
The test case should pass and allow for drive enumeration to occur for MSBuild to properly log a warning when a drive enumerating wildcard is detected.

### Actual Behavior
The test case throws an `UnauthorizedAccessException` and hangs.

### Analysis
This occurs for attempts to log a warning in ProjectItem_Tests.cs and ProjectItemInstance_Tests.cs as well.

### Versions & Configurations
Microsoft (R) Build Engine version 17.0.0+c9eb9dd64 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

17.0.0.52104

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes #7330
(plus one subtask of #8329)

Tests only change (no production code affected)

### Context
Drive enumeration integration tests need to simulate attempt to enumerate whole drive.
To prevent security and test runtime considerations - a dummy folder is created and mapped to a free letter to be offered to test as a drive for enumeration.

### Changes Made
Added utility for mapping drives and mounted to affected unit tests.

### CODE REVIEW

File: src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs [hunk 1]
```diff
@@ -806,6 +808,9 @@ public void ProjectGetterResultsInDriveEnumerationException(string unevaluatedIn
         [InlineData(@"z:\**\*.cs")]
         public void ProjectGetterResultsInWindowsDriveEnumerationWarning(string unevaluatedInclude)
         {
+            // let's create the mapped drive only once it's needed by any test, then let's reuse;
+            _mappedDrive ??= new DummyMappedDrive();
+            unevaluatedInclude = UpdatePathToMappedDrive(unevaluatedInclude, _mappedDrive.MappedDriveLetter);
```
Comment (vlada-shubina) [inline src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs:813]: nit: just wondering if we can request the specific drive instead, imo this makes it a bit more readable

```
using DummyMappedDrive mappedDrive = DriveMapping.GetDrive("z");

// test content


```

File: src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs [hunk 2]
```diff
@@ -901,10 +905,23 @@ public void ThrowExceptionUponProjectInstanceCreationFromDriveEnumeratingContent
...
             CleanContentsAndCreateProjectInstanceFromFileWithDriveEnumeratingWildcard(content, false);
         }
 
+        private static string UpdatePathToMappedDrive(string path, char driveLetter)
+        {
+            if (!string.IsNullOrEmpty(path) && path.StartsWith(driveLetter + ":", StringComparison.OrdinalIgnoreCase))
+            {
+                path = driveLetter + path.Substring(1);
+            }
+            return path;
```
Comment (Forgind) [inline src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs:922]: If path is empty, should we return "" or driveLetter + ":"? I'm curious if this could artificially make the drive enumeration tests pass.

**Review by Forgind — APPROVED** [review]

File: src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs (hunk 1 above)
Comment (JanKrivanek) [reply src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs]: The drive to request is unknown upfront - it needs to be untaken on the system.
So the `_mappedDrive` takes care about determining one that is free and using it.
Added `DriveMapping` is utility class that can be used for that purpose.

tl;dr;: Explicit assigning is intentionally hidden in this wrapper.

File: src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs (hunk 2 above)
Comment (JanKrivanek) [reply src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs:931]: Empty paths are valid test case scenarios (e.g. unspecified exclude pattern). So it intentionaly leaves unspecified or unrooted paths unaffected

File: src/UnitTests.Shared/DummyMappedDrive.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    private void ReleaseUnmanagedResources(bool disposing)
+    {
+        Exception? e = null;
+        if (Directory.Exists(_mappedPath))
+        {
+            try
+            {
+                Directory.Delete(_mappedPath, true);
```
Comment (ladipro) [inline src/UnitTests.Shared/DummyMappedDrive.cs:52]: nit: Wouldn't this always fail on non-Windows? The constructor sets `_mappedPath` but doesn't create anything. I understand the class is supposed to be used only on Windows but it contains explicit OS checks so it would be better if it had reasonable behavior everywhere.

File: src/UnitTests.Shared/DriveMapping.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    /// <summary>
+    /// Windows specific. Fetches path mapped under specific drive letter.
+    /// </summary>
+    /// <param name="letter">Drive letter.</param>
+    /// <returns>Path mapped under specified letter. Empty string if mapping not found.</returns>
+    [SupportedOSPlatform("windows")]
+    public static string GetDriveMapping(char letter)
+    {
+        // since this is just for test purposes - let's not overcomplicate with long paths support
```
Comment (ladipro) [inline src/UnitTests.Shared/DriveMapping.cs:56]: Wait, so if I have a drive mapped to a long path, unit tests will fail on my computer? I would vote for this to be more robust. Should be easy because we don't need the path, we just need to know if the mapping exists.

File: src/UnitTests.Shared/DriveMapping.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            }
+            NativeMethodsShared.ThrowExceptionForErrorCode(err);
+        }
+        // Translate from the native path semantic - starting with '\??\'
+        return new string(buffer, 4, buffer.Length - 4);
+    }
+
+    private static string ToDeviceName(char letter)
+    {
+        return new string(char.ToUpper(letter), 1) + ":";
```
Comment (ladipro) [inline src/UnitTests.Shared/DriveMapping.cs:74]: nit:
```suggestion
        return $"{char.ToUpper(letter)}:";
```

**Review by ladipro — COMMENTED** [review]
Looks great, leaving a couple of comments.

**Comment by ladipro** [comment]
We should also fix the Unix version of these tests. `ProjectGetterResultsInUnixDriveEnumerationWarning` accounts for almost a minute of test run-time 😢 

image

File: src/UnitTests.Shared/DummyMappedDrive.cs (hunk 3 above)
Comment (JanKrivanek) [reply src/UnitTests.Shared/DummyMappedDrive.cs:52]: It's wrapped in existence check - does that feel sufficient?

**Comment by JanKrivanek** [comment]
Added for unix long run and skipping for now: https://github.com/dotnet/msbuild/issues/8373

File: src/UnitTests.Shared/DummyMappedDrive.cs (hunk 3 above)
Comment (ladipro) [reply src/UnitTests.Shared/DummyMappedDrive.cs:52]: It does! I must have missed the check, apologies.

**Review by ladipro — APPROVED** [review]

File: src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs (hunk 1 above)
Comment (vlada-shubina) [reply src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs]: in this case, using placeholder in test data might be more readable, and then it is replaced by `DriveMapping`.
e.g: `"%DRIVE%:\**\*.cs"`

imo, for test purposes picking up 'z' or other letter at the end of abc should work.

Comment (JanKrivanek) [reply src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs]: I like the `%DRIVE%:` suggestion! - updated.

I wouldn't use any fixed drive letter though (ppl tend to map shares to various  drive letters). The dynamic drive letter allocation logic isn't complicated - so 'd keep it.
